x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/TopNOperator.java

```diff
@@ -264,12 +282,14 @@ public String describe() {
 
     private final BlockFactory blockFactory;
     private final CircuitBreaker breaker;
-    private final Queue inputQueue;
+    private final Map<String, Queue> inputQueues;
```

Comment:
I'm a bit worried about memory tracking for the keys.

Generally we use `ObjectArray` to keep things like the `Queue`s and we use `BytesRefHash` to assign the ids.

If we did that it'd read very very very similarly to an aggregation. In aggs-land we separate the hashing implementation from the "collecting" one. I wonder if we'd be better off using the aggs infrastructure in this case. No idea how we migrate from one to the other, but we have a `TopBytesRefAggregator`. And we could try to use that *somehow*. If we did that we'd get nice handling for the hashing for free. We're quite good at those hash keys.

Reply:
It is. It has the advantage of applying all of the grouping code we worked hard on for aggs so you don't have to hand build something on your side. Maybe you'd be better off just using the `BlockHash.build` infrastructure here. It'd be quite compatible with the ObjectArray stuff. I'm just worried that if you did it you'd find yourself building most of the `HashAggregatorOperator`.

I can have a little look around this morning though.

Reply:
I had a look and think it's probably not worth trying to migrate the whole thing to an agg. Aggs encode their intermediate representation as a single dense column. TopN builds the result columns. They aren't incompatible, but it'd be tricky to migrate.

I suppose what I suggest instead is to use `BlockHash` to turn the grouping keys into an integer and then `ObjectArray` to contain the heaps. In that case `BlockHash` is going to do a good job managing memory, breaking if it uses too much, and cleaning up. It also has support for non-string keys and complex sets of keys. And will absorb any optimizations we make in the future.